### ISSUE TITLE
calling `WindowDesc::set_position` has no effect on GTK when run on wayland

### ISSUE THREAD

**Issue by arifd** [linebender/druid#1757]
The window always opens in the same part (close to top left) of my monitor every time, no matter what values for `Point` i feed it.

This is likely also the reason that `Dropdown` from the widget-nursery is broken, as they too exhibit the same behavior and use `.set_position()` to place their dropdown menu.

**Comment by jneem** [comment]
Did this ever work on your system, or is it a regression?

**Comment by arifd** [comment]
First time using it, so I can't tell you.

My git-fu is not very strong, but my computer is reasonably enough, and I think with git, you can actually go into the past and compile for previous versions, so if you can tell me what to throw into the terminal, I'll be happy to do any testing / answer this question for you / etc

**Comment by jneem** [comment]
So I'd suggest working with the druid-widget-nursery repository first (and then we can try to nail down a druid commit later). I ran `git log src/dropdown.rs` to find that the first commit with dropdown support is `6c3e3512`. So first, checkout that commit to see if it worked then:

```sh
$ git checkout 6c3e3512
$ cargo run --example dropdown
```

If that one works, then you can use `git bisect` to find where it broke.

```sh
$ git bisect start
$ git bisect bad master # because the current version is broken
$ git bisect good 6c3e3512 # because the old version worked
```

Then git will automatically check out a commit for you. You test that commit, and type either `git bisect bad` or `git bisect good` depending on whether it worked or not. (If you're not sure, for example because that commit failed to compile, do `git bisect skip` instead.)

**Comment by arifd** [comment]
Hi @jneem, thanks for that, git continues to impress me.

Well, I can confirm that
```
$ git checkout 6c3e3512
$ cargo run --example dropdown
```
produces a broken example also, although the behaviour is ever so slightly different.

- In `git checkout master`
    - If I never move the parent window
         - the menu appears to open inside the window (don't know if my design or by coincidence (- looks like correct behaviour)), but since clicking anywhere else (including the parent) doesn't close the dropdown menu, i can click the button to open up more menus and they tile one after the other.
    - If I move the parent
        - all dropdown menus open at that familiar top left coordinate
     
- In `git checkout 6c3e3512`
     - both moving the parent or not, dropdowns always exhibit the same behaviour, that is, they open up (and tile) at that top left position

**Comment by jneem** [comment]
Ok, thanks for checking. One more thing: apparently Ubuntu 21.04 uses wayland by default; does anything change if you switch to x11?

**Comment by arifd** [comment]
Yes! logging in with `ubuntu under xorg`... `.set_position()` now works!

And the dropdown menu appears inside the parent window, however if you move the parent, the dropdown does not follow along.

**Comment by jneem** [comment]
Got it, thanks! I still don't know how to fix it, but at least this narrows it down. I'm guessing that the thing with the parent window moving is an issue with the dropdown code in druid-widget-nursery -- if it wants to move with the parent window, it should implement that manually.

**Comment by jneem** [comment]
A bit of googling suggests that we need to call [`Window::set_transient_for`](https://valadoc.org/gtk+-3.0/Gtk.Window.set_transient_for.html). In particular, I think this means that `druid-shell` needs some explicit notion of "subwindow"?

**Comment by emi2k01** [comment]
Global coordinates are not a thing on Wayland, as such, positioning of top-level windows is not possible under Wayland by design

[1]: https://gitlab.freedesktop.org/wayland/wayland/-/issues/183

**Comment by jneem** [comment]
Right, but my understanding (correct me if I'm wrong) is that wayland *does* support setting the position of a subwindow relative to its parent. As far as things like dropdowns and tooltips go, this should be enough. If I did my googling correctly (dubious), the `set_transient_for` function is the one we need, and the remaining issue is to figure out how to represent the subwindow/parent window relationship in druid, which doesn't currently have such a thing.

**Comment by JAicewizard** [comment]
The notation of setting a position does indeed not exist in GTK4/wayland, and is only possible under X11, the best you can do is set it relative to the main window, indeed using `set_transient_for`. I think this should be combined with `https://github.com/linebender/druid/issues/1824` (maybe that is the point of that issue?) to redo the window hierarchy in druid. I dont think you can go beyond 1 level deep in GTK4, so probably not in wayland either, but I could be wrong about that.

**Comment by JAicewizard** [comment]
Together with my proposal to remove `set_level` on non-builders in #1824 I would like to propose a change to the `WindowLevel` enum for 8.0:

```
pub enum WindowLevel {
    /// A top level app window.
    AppWindow,
    /// A window that should stay above app windows - like a tooltip
    Tooltip(Box<platform::WindowHandle>),
    /// A user interface element such as a dropdown menu or combo box
    DropDown(Box<platform::WindowHandle>),
    /// A modal dialog
    Modal(Box<platform::WindowHandle>),
}
```

This is useful mostly for GTK, and Wayland, but possibly also x11 and maybe other platforms.

This would give each subwindow, a parent window. This gives us a couple of things. We can use `set_transient_for` in GTK and similar stuff in other lineup backends. With this, we can make the coordinate system of the subwindows all relative to their parent. This not only makes sense from a windowing POV, but also allows us to correctly position under wayland solving this issue by improving GTK4/wayland compatibility.

Other backends will have to adapt their positioning code to be relative, and this would be a breaking chance in both the enum and `set_position`.

Ideally it would be impossible to have a tooltip or dropdown as a parent window. Somehow incorporating this into the type system seems hard, and its a bit wayland-centric.

**Comment by jneem** [comment]
I think this is fixed by #1919? I'm not in a position to easily test it.

**Comment by arifd** [comment]
Hooray! Can confirm now working under Wayland.
